File: 2764-maximum-number-of-fish-in-a-grid/maximum-number-of-fish-in-a-grid.java

class Solution {
    public int findMaxFish(int[][] grid) {
        int m = grid.length;
        if (m == 0) return 0;
        int n = grid[0].length;
        boolean[][] visited = new boolean[m][n];
        int maxFish = 0;
        int[][] directions = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
        
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (grid[i][j] > 0 && !visited[i][j]) {
                    Queue<int[]> queue = new LinkedList<>();
                    queue.add(new int[]{i, j});
                    visited[i][j] = true;
                    int currentSum = grid[i][j];
                    
                    while (!queue.isEmpty()) {
                        int[] cell = queue.poll();
                        for (int[] dir : directions) {
                            int r = cell[0] + dir[0];
                            int c = cell[1] + dir[1];
                            if (r >= 0 && r < m && c >= 0 && c < n && !visited[r][c] && grid[r][c] > 0) {
                                visited[r][c] = true;
                                currentSum += grid[r][c];
                                queue.add(new int[]{r, c});
                            }
                        }
                    }
                    maxFish = Math.max(maxFish, currentSum);
                }
            }
        }
        return maxFish;
    }
}

File: 2764-maximum-number-of-fish-in-a-grid/maximum-number-of-fish-in-a-grid.py
class Solution:
    def findMaxFish(self, grid: List[List[int]]) -> int:
        m = len(grid)
        n = len(grid[0]) if m > 0 else 0
        visited = [[False for _ in range(n)] for _ in range(m)]
        max_fish = 0
        directions = [(0, 1), (1, 0), (0, -1), (-1, 0)]
        
        for i in range(m):
            for j in range(n):
                if grid[i][j] > 0 and not visited[i][j]:
                    # BFS initialization
                    q = deque()
                    q.append((i, j))
                    visited[i][j] = True
                    current_sum = grid[i][j]
                    
                    while q:
                        r, c = q.popleft()
                        for dr, dc in directions:
                            nr, nc = r + dr, c + dc
                            if 0 <= nr < m and 0 <= nc < n:
                                if grid[nr][nc] > 0 and not visited[nr][nc]:
                                    visited[nr][nc] = True
                                    current_sum += grid[nr][nc]
                                    q.append((nr, nc))
                    max_fish = max(max_fish, current_sum)
        
        return max_fish